pkg/private/tar/build_tar.py

```diff
@@ -434,11 +451,11 @@ def main():
       compressor = options.compressor,
       default_mtime=default_mtime) as output:
 
-    def file_attributes(filename):
+    def file_attributes(filename, with_default_mode=True):
```

Comment:
Why is with_default_mode needed?  Having it, and only for mode rather than the rest makes this method a little more complex to understand.

Reply:
Without `with_default_mode`, I can't tell the difference between a value that's intentionally set and a value that's falling back to the default. This is important because file_attributes falls back to the default mode (as passed by the end-user caller); it's pretty common to set the default mode to something like 0644, which is a fine default for files but a terrible default for directories. `with_default_mode` allows me to have different defaults for directories and files (without a breaking change)

Reply:
I said "without a breaking change" but I guess I should note that there's a non-zero chance of a breaking change in this; before, if someone set `modes` to a file or a directory in a TreeArtifact, it would have been silently ignored before and now it would be honored. It's somewhat unlikely that anyone would actually be bitten by this (I noticed the bug when I tried to do exactly this, and noticed it didn't work) but it'd probably need to be called out.